exercises/bowling/example.rb

```diff
@@ -0,0 +1,66 @@
+class Game
+  class InvalidRoll < StandardError; end
+  class GameInProgress < StandardError; end
+  class GameHasEnded < StandardError; end
+
+  VERSION = 1
+  TOTAL_FRAMES = 10
+  MINIMUM_ROLL_SCORE = 0
+  MAXIMUM_ROLL_SCORE = 10
+
+  def initialize
+    @frame_scores = []
+    @frame_buffer = []
+  end
+
+  def roll(pins)
+    validate_roll(pins)
+    @frame_buffer << pins
+    adjust_frames
+  end
+
+  def score
+    raise GameInProgress unless @frame_scores.size == TOTAL_FRAMES
+    @frame_scores.reduce(:+)
+  end
+
+  private
+
+  def validate_roll(pins)
+    raise InvalidRoll if pins < MINIMUM_ROLL_SCORE
+    raise InvalidRoll if pins > MAXIMUM_ROLL_SCORE
+
+    if @frame_buffer.size == 1 && !strike_frame?
+      raise InvalidRoll if @frame_buffer[0] + pins > MAXIMUM_ROLL_SCORE
+    end
+
+    raise GameHasEnded if @frame_scores.size == TOTAL_FRAMES
+  end
+
+  def strike_frame?
+    @frame_buffer[0] == MAXIMUM_ROLL_SCORE
+  end
+
+  def spare_frame?
+    @frame_buffer[0] + @frame_buffer[1] == MAXIMUM_ROLL_SCORE
+  end
+
+  def open_frame?
+    !strike_frame? && !spare_frame?
+  end
+
+  def adjust_frames
+    if @frame_buffer.size == 3 && strike_frame?
+      @frame_scores << @frame_buffer.reduce(:+)
+      @frame_buffer.shift
+    elsif @frame_buffer.size == 3 && spare_frame?
+      @frame_scores << @frame_buffer.reduce(:+)
+      @frame_buffer.shift(2)
+    end
+
+    if @frame_scores.size < 10 && @frame_buffer.size == 2 && open_frame?
+      @frame_scores << @frame_buffer.reduce(:+)
+      @frame_buffer.shift(2)
+    end
+  end
```

Comment:
This is example. so doesn't necessarily have to be exemplary.

But... If these two sections were named based on what they are doing or what the specific condition is that this happens in, it may be easier to read/understand.

```ruby
  def adjust_frames
    if @frame_buffer.size == 3 && (strike_frame? || spare_frame?)
      @frame_scores << @frame_buffer.reduce(:+)
      @frame_buffer.shift(strike_frame? ? 1 : 2)
    end

    if @frame_scores.size < 10 && @frame_buffer.size == 2 && open_frame?
      @frame_scores << @frame_buffer.reduce(:+)
      @frame_buffer.shift(2)
    end
  end
```